src/ngx_http_lua_control.c

```diff
@@ -239,8 +243,17 @@ ngx_http_lua_ngx_redirect(lua_State *L)
                           "the headers");
     }
 
-    if (ngx_http_lua_check_unsafe_string(r, p, len, "redirect uri") != NGX_OK) {
-        return luaL_error(L, "attempt to set unsafe redirect uri");
+    if (ngx_http_lua_check_unsafe_uri_bytes(r, p, len, &byte) != NGX_OK) {
+        buf_len = ngx_http_lua_escape_log(NULL, p, len);
+        buf = ngx_palloc(r->pool, buf_len + 1);
+        if (buf == NULL) {
+            return NGX_ERROR;
+        }
+
+        ngx_http_lua_escape_log(buf, p, len);
+        buf[buf_len] = '\0';
+        sprintf(byte_hex, "%02x", byte);
+        return luaL_error(L, "unsafe byte \"0x%s\" in \"%s\"", byte_hex, buf);
```

Comment:
according to https://www.lua.org/manual/5.1/manual.html#lua_pushfstring.
“%x” is not allowed in luaL_xxx api although luajit can accpet it.

The conversion specifiers are quite restricted. There are no flags, widths, or precisions. The conversion specifiers can only be '%%' (inserts a '%' in the string), '%s' (inserts a zero-terminated string, with no size restrictions), '%f' (inserts a lua_Number), '%p' (inserts a pointer as a hexadecimal numeral), '%d' (inserts an int), and '%c' (inserts an int as a character).